framework.py

```diff
@@ -9,16 +9,150 @@
+        """
         self.log_debug("%s: Initializing..." % self)
+        self._stop_cleanup = False
+        self._bg_cleanup_thread = None
+        self._post_old_data_cleanup()
     
     def destroy_framework(self):
+        """
+        Destroy this framework.
+
+        If an old cached data cleanup was posted in the background, stop it
+        immediately.
+        """
         self.log_debug("%s: Destroying..." % self)
+        # Please note that we are modifying a member which is read in another
+        # thread which should be fine in Python with the GIL protecting its access.
+        self._stop_cleanup = True
+        if self._bg_cleanup_thread:
+            if self._bg_cleanup_thread.isAlive():
+                # If the clean up is not completed yet, log why we are waiting.
+                self.log_info("Waiting for old data clean up to complete...")
+            self._bg_cleanup_thread.join()
     
-    
+    def _post_old_data_cleanup(self):
+        """
+        Post a cleanup of old cached data in the background.
+        """
+        try:
+            self.log_debug(
+                "Posting old cached data clean up..."
+            )
+            self._stop_cleanup = False
+            # Qt might not be yet available at this stage (e.g. in tk-desktop),
+            # so we can't use a background task manager or a QThread, use instead
+            # regular Python Thread to post the clean up in the background.
+            self._bg_cleanup_thread = threading.Thread(
+                target=self._remove_old_cached_data,
+                name="%s Clean Up" % self.name
+            )
+            self._bg_cleanup_thread.start()
+        except Exception as e:
+            self.log_warning("Unable to post data clean up: %s" % e)
+
+    def _remove_old_cached_data(self, grace_period=7):
+        """
+        Remove old data files cached by this bundle.
+
+        A file is considered old if it was not modified in the last number of days
+        specified by the `grace_period` value, which must be at least 1 (one day).
+
+        It is the responsability of the implementation to ensure that modification
+        times for the files which should be kept are recent.
+        Typically, when re-using a cached file, the bundle should use
+        `os.utime(cached_file_path, None)` to update the modification time to the
+        current time.
+
+        If some files should never be deleted, their name should be added to the
+        `_ALWAYS_KEEP_CACHED_FILES` class member.
+
+        :param int grace_period: The number of days files without any modification
+                                 should be kept around before being deleted.
+        :raises: ValueError if the grace_period is smaller than 1.
+        """
+        if grace_period < 1:
+            raise ValueError(
+                "Invalid grace period value %d, it must be a least 1" % grace_period
+            )
+        self.logger.debug("Starting old data cleanup...")
```

Comment:
please include the grace period date in this message - that seems useful, maybe also explain what it will do. Something like "Cleaning all files with a modofication date older than xxxx in location ~/Library/caches/.../thumbnails"

Reply:
```
Cleaning all files with a modification date older than 2017-12-18
18:32:16.358171 under locations
/Users/steph/Library/Caches/Shotgun/couchlab/site/fw-shotgunutils/sg,
/Users/steph/Library/Caches/Shotgun/couchlab/site/fw-shotgunutils/sg_nav,
/Users/steph/Library/Caches/Shotgun/couchlab/site/fw-shotgunutils/thumbs,
/Users/steph/Library/Caches/Shotgun/couchlab/p309c171/fw-shotgunutils/sg,
/Users/steph/Library/Caches/Shotgun/couchlab/p309c171/fw-shotgunutils/sg_nav,
/Users/steph/Library/Caches/Shotgun/couchlab/p309c171/fw-shotgunutils/thumbs
```